src/blueapi/service/runner.py

```diff
@@ -152,6 +152,36 @@ def _rpc(
         )
 
 
+def is_valid_return(value: Any, expected_type: type[T]) -> bool:
+    if expected_type is Any:
+        return True
+    if expected_type is type(None):
+        return value is None
+    origin = get_origin(expected_type)
+    if origin is None or origin is Union:  # works for Python >= 3.10
+        return isinstance(value, expected_type)
+    args = get_args(expected_type)
+    if issubclass(origin, Mapping):
+        return isinstance(value, Mapping) and all(
+            is_valid_return(k, args[0]) and is_valid_return(v, args[1])
+            for k, v in value.items()
+        )
+    if origin is tuple:  # handle ellipses specially
+        assert isinstance(value, tuple)
+        if args[1] is Ellipsis:
+            args = (args[0],) * len(value)
+        else:
+            return all(
+                is_valid_return(inner, arg)
+                for inner, arg in zip(value, args, strict=True)
+            )
+    if issubclass(origin, Collection):  # list, set, etc.
+        return isinstance(value, Collection) and all(
+            is_valid_return(inner, args[0]) for inner in value
+        )
+    raise TypeError(f"Unknown origin for generic type {expected_type}")
```

Comment:
Could `pydantic.TypeAdapter` do all of this?

Reply:
If it can, I can't test until #449

Reply:
@callumforrester with Pydantic.TypeAdapter it's still got an issue with the same single case. My handling is equally as good ;)